### PR TITLE
Allow Jinja2 whitespace control chars in vars

### PR DESCRIPTION
**Pull request by netzvieh**

### CODE REVIEW

**Comment by netzvieh** [comment]
Fixes #499

**Comment by ssbarnea** [comment]
@webknjaz Can you have a look at this?

File: lib/ansiblelint/rules/VariableHasSpacesRule.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ class VariableHasSpacesRule(AnsibleLintRule):
     tags = ['formatting']
     version_added = 'v4.0.0'
 
-    bracket_regex = re.compile("{{[^{' ]|[^ '}]}}")
+    bracket_regex = re.compile("{{[^{' -]|[^ '}-]}}")
```
Comment (webknjaz) [inline lib/ansiblelint/rules/VariableHasSpacesRule.py]: please use raw-string literals for regexps
```suggestion
    bracket_regex = re.compile(r"{{[^{' -]|[^ '}-]}}")
```

File: test/TestVariableHasSpaces.py [hunk 2]
```diff
@@ -11,11 +11,16 @@
...
   debug:
     msg: "{{ '{{' }}"
-- name: jijna escaping allowed
+- name: jinja escaping allowed
   shell: docker info --format '{{ '{{' }}json .Swarm.LocalNodeState{{ '}}' }}' | tr -d '"'
+- name: jinja whitespace control allowed
+  debug:
+    msg: |
+      {{ good_format }}/
+      {{- good_format }}
```
Comment (webknjaz) [inline test/TestVariableHasSpaces.py]: extend this test with @ssbarnea's one

File: lib/ansiblelint/rules/VariableHasSpacesRule.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansiblelint/rules/VariableHasSpacesRule.py]: I'm not sure why this regexp does this check. Can anyone explain it to me? It seems wrong (even the original one)

File: test/TestVariableHasSpaces.py (hunk 2 above)
Comment (ssbarnea) [reply test/TestVariableHasSpaces.py]: @netzvieh I hope you can make the small fix that was requested as I want to help merging this bugfix.

File: lib/ansiblelint/rules/VariableHasSpacesRule.py (hunk 1 above)
Comment (netzvieh) [reply lib/ansiblelint/rules/VariableHasSpacesRule.py]: As I understand we want to match everything that's starting with `{{` not followed by space or hyphen. Also we want to allow `{{ '{{' }}` so we need to allow single quotes after `{{` aswell. Same thing for `}}`. I'm not sure why we wouldn't want to match `{{{` or `}}}` though.

Maybe @awcrosby can tell us why it's needed?

Problem is, this actually allows doing `{{'asdas'}}` as `{{'` and `'}}` are not matched. Same for `{{-asassa-}}`. But the regex to capture that would be something like: `({{-(?! ))|({{'(?! ))|({{[^ '-])|([^' -]}})|((?<=[^ ])-}})|((?<=[^ ])'}})` and I'm not positive we want to do that.

I'm not sure if python does allow the regex to be split into multiple lines? Then we probably could introduce it with `(# comment)` to explain it. Basically like:

```
({{-(?! ))|(# match everything with {{- not followed by a space OR)
({{'(?! ))|(# match everything with {{' not followed by a space OR)
({{[^ '-])|(# match everything with {{ not followed by a space, single quotation mark or hyphen OR)
([^' -]}})|(# match everything with }} not preceded by a space, single quotation mark or hyphen OR)
((?<=[^ ])-}})|(# match everything with -}} not preceded by a space OR)
((?<=[^ ])'}})(# match everything with '}} not preceded by a space)
```

Comment (netzvieh) [reply lib/ansiblelint/rules/VariableHasSpacesRule.py]: It does work with `\` at EOL, though I'm still not sure, we want to do that.

File: test/TestVariableHasSpaces.py (hunk 2 above)
Comment (netzvieh) [reply test/TestVariableHasSpaces.py]: Added.

**Comment by awcrosby** [comment]
The allowing of `{{{` and `}}}` was added to the original rule in ansible-review before the rule was added to ansible-lint, I am not sure the purpose.

Thanks for pointing out this rule allows `{{'asdas'}}`, this may be ok for now as we are catching the most common bad formats, and limiting the number of false positives.

The PR looks good, I will merge this in - thanks for the contribution!

**Pull request merged**